File: kotlin/bc-ur/src/main/kotlin/com/blockchaincommons/bcur/MultipartDecoder.kt
package com.blockchaincommons.bcur

import com.blockchaincommons.dcbor.Cbor

/** Multipart UR decoder using fountain codes. */
class MultipartDecoder {
    private var urType: URType? = null
    private val decoder = FountainDecoder()

    /**
     * Receives a multipart UR string.
     *
     * Accepts both lowercase and uppercase input.
     *
     * @throws URException if the string is malformed or the type is inconsistent
     */
    fun receive(value: String) {
        val lower = value.lowercase()
        val decodedType = decodeType(lower)
        val current = urType
        if (current != null) {
            if (current != decodedType) {
                throw URException.UnexpectedType(current.value, decodedType.value)
            }
        } else {
            urType = decodedType
        }

        val (kind, data) = UREncoding.decode(lower)
        if (kind != UREncoding.Kind.MultiPart) {
            throw URException.NotSinglePart()
        }
        val part = FountainPart.fromCbor(data)
        decoder.receive(part)
    }

    /** Returns whether the decoder has received all fragments. */
    val isComplete: Boolean get() = decoder.isComplete

    /** Number of fragments that have been fully decoded. */
    val decodedFragmentCount: Int get() = decoder.decodedCount

    /** Total number of fragments needed. Returns 0 before the first part is received. */
    val expectedFragmentCount: Int get() = decoder.expectedCount

    /** Set of fragment indexes that have been fully decoded. */
    val decodedFragmentIndexes: Set<Int> get() = decoder.decodedIndexes

    /** Partial progress credit from buffered mixed-degree parts. */
    val bufferContribution: Double get() = decoder.bufferContribution

    /** Returns the decoded UR if complete, or null. */
    fun message(): UR? {
        val data = decoder.message() ?: return null
        val cbor = Cbor.tryFromData(data)
        val type = urType!!
        return UR(type, cbor)
    }

    private fun decodeType(urString: String): URType {
        if (!urString.startsWith("ur:")) {
            throw URException.InvalidScheme()
        }
        val withoutScheme = urString.removePrefix("ur:")
        val firstComponent = withoutScheme.split('/').firstOrNull()
            ?: throw URException.InvalidType()
        return URType(firstComponent)
    }
}

File: swift/BCUR/Sources/BCUR/MultipartDecoder.swift
import Foundation
import DCBOR

/// A multipart UR decoder that reconstructs a full UR from received parts.
public final class MultipartDecoder {
    private var urType: URType?
    private var decoder = FountainDecoder()

    /// Creates an empty multipart decoder.
    public init() { }

    /// Receives a multipart UR part string.
    ///
    /// Input is normalized to lowercase to support uppercase QR payloads.
    public func receive(_ value: String) throws {
        let normalized = value.lowercased()
        let decodedType = try decodeType(normalized)

        if let existing = urType {
            if existing != decodedType {
                throw URError.unexpectedType(existing.string, decodedType.string)
            }
        } else {
            urType = decodedType
        }

        let decoded: (UREncoding.Kind, [UInt8])
        do {
            decoded = try UREncoding.decode(normalized)
        } catch let error as URCodecError {
            throw URError(ur: error)
        }

        guard decoded.0 == .multiPart else {
            throw URError(ur: .notMultiPart)
        }

        let part: FountainPart
        do {
            part = try FountainPart.fromCbor(decoded.1)
        } catch let error as FountainError {
            throw URError(fountain: error)
        }

        do {
            _ = try decoder.receive(part)
        } catch let error as FountainError {
            throw URError(fountain: error)
        }
    }

    /// Indicates whether all fragments have been received.
    public var isComplete: Bool {
        decoder.complete
    }

    /// Returns the reconstructed UR if complete, else `nil`.
    public func message() throws -> UR? {
        let messageData: [UInt8]?
        do {
            messageData = try decoder.message()
        } catch let error as FountainError {
            throw URError(fountain: error)
        }

        guard let messageData else {
            return nil
        }

        guard let urType else {
            throw URError.invalidType
        }

        do {
            let cbor = try CBOR(Data(messageData))
            return UR(urType, cbor)
        } catch {
            throw URError.fromCBORError(error)
        }
    }

    private func decodeType(_ urString: String) throws -> URType {
        guard urString.hasPrefix("ur:") else {
            throw URError.invalidScheme
        }

        let withoutScheme = urString.dropFirst(3)
        let firstComponent = String(withoutScheme.prefix { $0 != "/" })
        return try URType(firstComponent)
    }
}
